salt/_pillar/velum.py

```diff
@@ -0,0 +1,102 @@
+# -*- coding: utf-8 -*-
+'''
+A module that adds data to the Pillar structure retrieved by an http request made to Velum
+
+
+Configuring the VELUM ext_pillar
+====================================
+
+Set the following Salt config to setup Foreman as external pillar source:
+
+.. code-block:: json
+
+  ext_pillar:
+    - velum:
+        - url: https://velum.example.com/internal-api/v1/pillar?version=1
+        - ca_bundle: /some/bundle.crt
+        - username: username
+        - password: secret
+
+Module Documentation
+====================
+'''
+
+# Import python libs
+from __future__ import absolute_import
+import logging
+
+# Import Salt libs
+import salt.ext.six as six
+
+import json
+import hashlib
+import tempfile
+
+pillar_sha256 = None
+pillar_cache = None
+cached_username = None
+cached_password = None
+
+def ext_pillar(minion_id,
+               pillar,  # pylint: disable=W0613
+               url=None,
+               ca_bundle=None,
+               username=None,
+               password=None):
+    '''
+    Read pillar data from HTTP response.
+
+    :param url String to make request
+    :param ca_bundle Path to CA bundle
+    :param username Username for basic-auth
+    :param password Password for basic-auth
+    :returns dict with pillar data to add
+    :returns empty if error
+    '''
+
+    global pillar_sha256
+    global pillar_cache
+    global cached_username
+    global cached_password
+
+    # Set up logging
+    log = logging.getLogger(__name__)
+
+    if username is None:
+        if cached_username is None:
+            with open('/var/lib/misc/infra-secrets/velum-internal-api-username', 'r') as f:
+                cached_username = f.read().strip()
+        username = cached_username
+
+    if password is None:
+        if cached_password is None:
+            with open('/var/lib/misc/infra-secrets/velum-internal-api-password', 'r') as f:
+                cached_password = f.read().strip()
+        password = cached_password
+
+    data = __salt__['http.query'](url=url,
+                                  ca_bundle=ca_bundle,
+                                  username=username,
+                                  password=password,
+                                  decode=True,
+                                  decode_type='json')
+
+    if 'dict' in data:
+        current_sha256 = hashlib.sha256(json.dumps(data['dict']))
+        if pillar_sha256 is None or pillar_sha256 != current_sha256:
+            if pillar_cache:
+                pillar_cache.close()
+            pillar_cache = tempfile.TemporaryFile()
+            pillar_cache.write(json.dumps(data['dict']))
+            pillar_sha256 = current_sha256
+        return data['dict']
+    elif pillar_cache:
+        pillar_cache.seek(0)
```

Comment:
Couldn't we just keep an in-memory copy of the pillar (a `pillar_cache_contents` global) instead of trying to load it from the file each time `dict is None`?

Reply:
Probably yes, will review this. We will lost the file anyways if the container or the host is restarted so it probably makes sense to directly cache it in memory...